var/spack/repos/builtin/packages/py-deepsig-biocomp/package.py

```diff
@@ -0,0 +1,36 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+from spack.package import *
+
+
+class PyDeepsigBiocomp(PythonPackage):
+    """DeepSig - Predictor of signal peptides
+    in proteins based on deep learning"""
+
+    homepage = "https://deepsig.biocomp.unibo.it"
+
+    url = "https://github.com/BolognaBiocomp/deepsig/archive/refs/tags/v1.2.5.tar.gz"
+
+    version("1.2.5", sha256="e954b815d63c221c564c7d3fe27123d7cd2c39b191d6107369ab095d506496e0")
+
+    depends_on("python@3.8", type=("build", "run"))
+    depends_on("py-setuptools", type="build")
+    depends_on("py-biopython@1.78:", type=("build", "run"))
+    depends_on("py-keras@2.4.3", type=("build", "run"))
+    depends_on("py-tensorflow@2.2.0", type=("build", "run"))
+    depends_on("py-tensorboard", type=("build", "run"))
```

Comment:
Where does the TF version and TB dep come from? I see a TF version in the `environment.yml` although that isn't used by setuptools.

Reply:
yes, the documentation is a little scattershot. I don't know exactly what is correct, but the requirements.txt and bioconda recipes specify v2.2.0 exactly:

https://github.com/bioconda/bioconda-recipes/tree/master/recipes/deepsig
https://github.com/BolognaBiocomp/deepsig/blob/v1.2.5/requirements.txt

The TB dep came from the fact that when I built it without TB, I got a "Tensorboard required" error while testing the install. I dug into the TF package and the relationship between TF and TB deps just confused me and it seemed way easier to just require it for this package instead of dealing with circular TF dependencies.

Reply:
Can you try building with a modern version of TF and see if it works? I generally don't trust bioconda or requirements.txt since pip uses neither.

Reply:
When I remove the TF version requirement and TB requirement entirely, it builds with TF 2.8.3 (which also installs TB 2.8). However, when I tested the program I get the following error:

```
[lukedt@fedora]~/Documents/deepsig% deepsig --help
Traceback (most recent call last):
  File "/home/lukedt/Documents/GitHub/spack/opt/spack/linux-fedora36-zen3/gcc-12.2.1/py-keras-2.4.3-krozj2fhiynqyzh345qxrgw4mh3772dm/lib/python3.8/site-packages/keras/__init__.py", line 3, in <module>
    from tensorflow.keras.layers.experimental.preprocessing import RandomRotation
ModuleNotFoundError: No module named 'tensorflow.keras'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/lukedt/Documents/GitHub/spack/opt/spack/linux-fedora36-zen3/gcc-12.2.1/py-deepsig-biocomp-1.2.5-m4775sv6rntfq77lssxyn7crrcbyvuut/bin/deepsig", line 7, in <module>
    from deepsig.deepsig import main
  File "/home/lukedt/Documents/GitHub/spack/opt/spack/linux-fedora36-zen3/gcc-12.2.1/py-deepsig-biocomp-1.2.5-m4775sv6rntfq77lssxyn7crrcbyvuut/lib/python3.8/site-packages/deepsig/deepsig.py", line 15, in <module>
    from deepsig.helpers import setUpTFCPU, readdata, printDate, write_gff_output, get_json_output, detectsp, predictsp
  File "/home/lukedt/Documents/GitHub/spack/opt/spack/linux-fedora36-zen3/gcc-12.2.1/py-deepsig-biocomp-1.2.5-m4775sv6rntfq77lssxyn7crrcbyvuut/lib/python3.8/site-packages/deepsig/helpers.py", line 13, in <module>
    from keras.preprocessing.sequence import pad_sequences
  File "/home/lukedt/Documents/GitHub/spack/opt/spack/linux-fedora36-zen3/gcc-12.2.1/py-keras-2.4.3-krozj2fhiynqyzh345qxrgw4mh3772dm/lib/python3.8/site-packages/keras/__init__.py", line 5, in <module>
    raise ImportError(
ImportError: Keras requires TensorFlow 2.2 or higher. Install TensorFlow via `pip install tensorflow`
```

Here's the concretized spec:

```
Input spec
--------------------------------
py-deepsig-biocomp

Concretized
--------------------------------
py-deepsig-biocomp@1.2.5%gcc@12.2.1 arch=linux-fedora36-zen3
    ^py-biopython@1.79%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-numpy@1.23.3%gcc@12.2.1+blas+lapack patches=873745d arch=linux-fedora36-zen3
            ^openblas@0.3.21%gcc@12.2.1~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared symbol_suffix=none threads=none arch=linux-fedora36-zen3
                ^perl@5.34.1%gcc@12.2.1+cpanm+shared+threads arch=linux-fedora36-zen3
                    ^berkeley-db@18.1.40%gcc@12.2.1+cxx~docs+stl patches=26090f4,b231fcc arch=linux-fedora36-zen3
                    ^bzip2@1.0.8%gcc@12.2.1~debug~pic+shared arch=linux-fedora36-zen3
                        ^diffutils@3.8%gcc@12.2.1 arch=linux-fedora36-zen3
                            ^libiconv@1.16%gcc@12.2.1 libs=shared,static arch=linux-fedora36-zen3
                    ^gdbm@1.19%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^readline@8.1.2%gcc@12.2.1 arch=linux-fedora36-zen3
                            ^ncurses@6.3%gcc@12.2.1~symlinks+termlib abi=none arch=linux-fedora36-zen3
                                ^pkgconf@1.8.0%gcc@12.2.1 arch=linux-fedora36-zen3
                    ^zlib@1.2.12%gcc@12.2.1+optimize+pic+shared patches=0d38234 arch=linux-fedora36-zen3
            ^py-cython@0.29.32%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-pip@22.2.2%gcc@12.2.1 arch=linux-fedora36-zen3
                    ^python@3.8.13%gcc@12.2.1+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-fedora36-zen3
                        ^expat@2.4.8%gcc@12.2.1+libbsd arch=linux-fedora36-zen3
                            ^libbsd@0.11.5%gcc@12.2.1 arch=linux-fedora36-zen3
                                ^libmd@1.0.4%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^gettext@0.21%gcc@12.2.1+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-fedora36-zen3
                            ^libxml2@2.10.1%gcc@12.2.1~python arch=linux-fedora36-zen3
                                ^xz@5.2.5%gcc@12.2.1~pic libs=shared,static arch=linux-fedora36-zen3
                            ^tar@1.34%gcc@12.2.1 zip=pigz arch=linux-fedora36-zen3
                                ^pigz@2.7%gcc@12.2.1 arch=linux-fedora36-zen3
                                ^zstd@1.5.2%gcc@12.2.1+programs compression=none libs=shared,static arch=linux-fedora36-zen3
                        ^libffi@3.4.2%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^openssl@1.1.1q%gcc@12.2.1~docs~shared certs=mozilla patches=3fdcf2d arch=linux-fedora36-zen3
                            ^ca-certificates-mozilla@2022-07-19%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^sqlite@3.39.2%gcc@12.2.1+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-fedora36-zen3
                        ^util-linux-uuid@2.37.4%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-setuptools@59.4.0%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-wheel@0.37.1%gcc@12.2.1 arch=linux-fedora36-zen3
    ^py-keras@2.4.3%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-h5py@3.7.0%gcc@12.2.1~mpi arch=linux-fedora36-zen3
            ^hdf5@1.12.2%gcc@12.2.1~cxx~fortran+hl~ipo~java~mpi+shared~szip~threadsafe+tools api=default build_type=RelWithDebInfo arch=linux-fedora36-zen3
                ^cmake@3.24.2%gcc@12.2.1~doc+ncurses+ownlibs~qt build_type=Release arch=linux-fedora36-zen3
            ^py-pkgconfig@1.5.5%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-poetry-core@1.2.0%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-pandas@1.5.0%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-bottleneck@1.3.5%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-versioneer@0.26%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-numexpr@2.8.3%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-packaging@21.3%gcc@12.2.1 arch=linux-fedora36-zen3
                    ^py-pyparsing@3.0.9%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^py-flit-core@3.7.1%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-python-dateutil@2.8.2%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-setuptools-scm@7.0.5%gcc@12.2.1+toml arch=linux-fedora36-zen3
                    ^py-tomli@2.0.1%gcc@12.2.1 arch=linux-fedora36-zen3
                    ^py-typing-extensions@4.3.0%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-six@1.16.0%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-pytz@2022.2.1%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-pillow@9.2.0%gcc@12.2.1~freetype~imagequant+jpeg~jpeg2000~lcms~raqm~tiff~webp~xcb+zlib arch=linux-fedora36-zen3
            ^libjpeg-turbo@2.1.3%gcc@12.2.1 arch=linux-fedora36-zen3
                ^nasm@2.15.05%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-pydot@1.4.2%gcc@12.2.1 arch=linux-fedora36-zen3
            ^graphviz@2.49.0%gcc@12.2.1~doc~expat~ghostscript~gtkplus~gts~java~libgd~pangocairo~poppler~qt~quartz~x arch=linux-fedora36-zen3
                ^autoconf@2.69%gcc@12.2.1 patches=35c4492,7793209,a49dd5b arch=linux-fedora36-zen3
                    ^m4@1.4.19%gcc@12.2.1+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-fedora36-zen3
                        ^libsigsegv@2.13%gcc@12.2.1 arch=linux-fedora36-zen3
                ^automake@1.16.5%gcc@12.2.1 arch=linux-fedora36-zen3
                ^bison@3.8.2%gcc@12.2.1 arch=linux-fedora36-zen3
                ^flex@2.6.3%gcc@12.2.1+lex~nls arch=linux-fedora36-zen3
                    ^findutils@4.9.0%gcc@12.2.1 patches=440b954 arch=linux-fedora36-zen3
                ^libtool@2.4.7%gcc@12.2.1 arch=linux-fedora36-zen3
                ^sed@4.8%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-pyyaml@6.0%gcc@12.2.1+libyaml arch=linux-fedora36-zen3
            ^libyaml@0.2.5%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-scipy@1.8.1%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-pybind11@2.8.1%gcc@12.2.1~ipo build_type=RelWithDebInfo arch=linux-fedora36-zen3
                ^ninja@1.11.0%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-pythran@0.11.0%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-beniget@0.4.1%gcc@12.2.1 arch=linux-fedora36-zen3
                    ^py-gast@0.5.3%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-ply@3.11%gcc@12.2.1 arch=linux-fedora36-zen3
    ^py-tensorflow@2.8.3%gcc@12.2.1~android~aws~computecpp~cuda~dynamic_kernels~gcp~gdr~hdfs~ignite~ios~jemalloc~kafka~mkl~monolithic~mpi~nccl~ngraph~numa~opencl~rocm~tensorrt~verbs~xla patches=2017b3e arch=linux-fedora36-zen3
        ^bazel@4.2.2%gcc@12.2.1+nodepfail patches=0f6940d,3e6448a,604423c,86735e5,ae01da1,b4dfb11,becacef,dbbf38b arch=linux-fedora36-zen3
            ^openjdk@11.0.15_10%gcc@12.2.1 arch=linux-fedora36-zen3
            ^zip@3.0%gcc@12.2.1 patches=14dc880,3bc30ba,5068e7c,51f48db,66ab4ce,a92fc4e,a95ed93,eb83fc8,f7d0bc4,fa8312c arch=linux-fedora36-zen3
        ^protobuf@3.17.3%gcc@12.2.1+shared build_type=Release arch=linux-fedora36-zen3
        ^py-absl-py@1.2.0%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-astunparse@1.6.3%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-flatbuffers@2.0.7%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-google-pasta@0.2.0%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-grpcio@1.48.1%gcc@12.2.1 arch=linux-fedora36-zen3
            ^c-ares@1.15.0%gcc@12.2.1~ipo build_type=RelWithDebInfo arch=linux-fedora36-zen3
            ^re2@2021-06-01%gcc@12.2.1~ipo+pic+shared build_type=RelWithDebInfo arch=linux-fedora36-zen3
        ^py-keras-preprocessing@1.1.2%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-libclang@14.0.6%gcc@12.2.1 arch=linux-fedora36-zen3
            ^llvm@14.0.6%gcc@12.2.1+clang+compiler-rt~cuda~flang+gold+internal_unwind~ipo+libcxx~link_llvm_dylib+lld+lldb+llvm_dylib~mlir+omp_as_runtime~omp_debug~omp_tsan+polly~python~split_dwarf~z3 build_type=Release patches=6379168,d85ef51,f920173 shlib_symbol_version=none targets=none version_suffix=none arch=linux-fedora36-zen3
                ^binutils@2.38%gcc@12.2.1~gas+gold~headers~interwork+ld~libiberty~lto+nls+plugins libs=shared,static arch=linux-fedora36-zen3
                ^hwloc@2.8.0%gcc@12.2.1~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm libs=shared,static arch=linux-fedora36-zen3
                    ^libpciaccess@0.16%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^util-macros@1.19.3%gcc@12.2.1 arch=linux-fedora36-zen3
                ^libedit@3.1-20210216%gcc@12.2.1 arch=linux-fedora36-zen3
                ^perl-data-dumper@2.173%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-opt-einsum@3.3.0%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-protobuf@3.19.4%gcc@12.2.1~cpp arch=linux-fedora36-zen3
        ^py-tensorboard@2.8.0%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-google-auth@2.11.0%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-cachetools@5.2.0%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-pyasn1-modules@0.2.8%gcc@12.2.1 arch=linux-fedora36-zen3
                    ^py-pyasn1@0.4.8%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-rsa@4.7.2%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-google-auth-oauthlib@0.4.6%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-requests-oauthlib@1.3.0%gcc@12.2.1 arch=linux-fedora36-zen3
                    ^py-oauthlib@3.1.0%gcc@12.2.1+extras arch=linux-fedora36-zen3
                        ^py-blinker@1.4%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^py-cryptography@38.0.1%gcc@12.2.1 arch=linux-fedora36-zen3
                            ^git@2.37.0%gcc@12.2.1+man+nls+perl+subtree~svn~tcltk arch=linux-fedora36-zen3
                                ^curl@7.85.0%gcc@12.2.1~gssapi~ldap~libidn2~librtmp~libssh~libssh2~nghttp2 libs=shared,static tls=openssl arch=linux-fedora36-zen3
                                ^libidn2@2.3.0%gcc@12.2.1 arch=linux-fedora36-zen3
                                    ^libunistring@0.9.10%gcc@12.2.1 arch=linux-fedora36-zen3
                                ^openssh@9.0p1%gcc@12.2.1+gssapi arch=linux-fedora36-zen3
                                    ^krb5@1.19.3%gcc@12.2.1+shared arch=linux-fedora36-zen3
                                ^pcre2@10.39%gcc@12.2.1~jit+multibyte arch=linux-fedora36-zen3
                            ^py-cffi@1.15.0%gcc@12.2.1 arch=linux-fedora36-zen3
                                ^py-pycparser@2.20%gcc@12.2.1 arch=linux-fedora36-zen3
                            ^py-setuptools-rust@1.2.0%gcc@12.2.1 arch=linux-fedora36-zen3
                                ^py-semantic-version@2.10.0%gcc@12.2.1 arch=linux-fedora36-zen3
                                ^rust@1.60.0%gcc@12.2.1+analysis+clippy~rls+rustfmt+src extra_targets=none arch=linux-fedora36-zen3
                                    ^gmake@4.3%gcc@12.2.1~guile+nls arch=linux-fedora36-zen3
                                        ^texinfo@6.5%gcc@12.2.1 patches=12f6edb,1732115 arch=linux-fedora36-zen3
                                    ^libgit2@1.3.1%gcc@12.2.1~curl~ipo+mmap+ssh build_type=RelWithDebInfo https=system arch=linux-fedora36-zen3
                                        ^libssh2@1.10.0%gcc@12.2.1~ipo+shared build_type=RelWithDebInfo crypto=openssl arch=linux-fedora36-zen3
                        ^py-pyjwt@2.4.0%gcc@12.2.1~crypto arch=linux-fedora36-zen3
                    ^py-requests@2.28.1%gcc@12.2.1~socks arch=linux-fedora36-zen3
                        ^py-certifi@2022.9.14%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^py-charset-normalizer@2.0.12%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^py-idna@3.4%gcc@12.2.1 arch=linux-fedora36-zen3
                        ^py-urllib3@1.26.12%gcc@12.2.1~brotli~secure~socks arch=linux-fedora36-zen3
            ^py-markdown@3.4.1%gcc@12.2.1 arch=linux-fedora36-zen3
                ^py-importlib-metadata@4.12.0%gcc@12.2.1 arch=linux-fedora36-zen3
                    ^py-zipp@3.8.1%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-tensorboard-data-server@0.6.1%gcc@12.2.1 patches=878bbd6 arch=linux-fedora36-zen3
            ^py-tensorboard-plugin-wit@1.8.1%gcc@12.2.1 arch=linux-fedora36-zen3
            ^py-werkzeug@2.0.2%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-termcolor@1.1.0%gcc@12.2.1 arch=linux-fedora36-zen3
        ^py-wrapt@1.13.3%gcc@12.2.1 arch=linux-fedora36-zen3
        ^swig@4.0.2%gcc@12.2.1 arch=linux-fedora36-zen3
            ^pcre@8.45%gcc@12.2.1~jit+multibyte+utf arch=linux-fedora36-zen3
```

Reply:
Hmm, that is very odd. Well, if it works when pinned, I'm fine with keeping it pinned.